File: javascript/src/queue/Supervisor.ts
import { BaseRunner } from './BaseRunner';
import { Configuration } from './Configuration';
import { sleep } from './utils';
import { writeFileSync, mkdirSync } from 'fs';
import path from 'path';
export class Supervisor extends BaseRunner {
  constructor(redisUrl: string, config: Configuration) {
    super(redisUrl, config);
    this.isMaster = false;
  }

  async totalTests() {
    await this.waitForMaster();
    return await this.client.get(this.key('total'));
  }

  async waitForWorkers() {
    await this.waitForMaster();
    let timeLeft = this.config.reportTimeout;
    let timeLeftWithNoWorkers = this.config.inactiveWorkersTimeout;
    while (
      !(await this.isExhausted()) &&
      timeLeft > 0 &&
      !(await this.maxTestsFailed()) &&
      timeLeftWithNoWorkers > 0
    ) {
      timeLeft -= 1;
      await sleep(1000);

      if (await this.workersAreActive()) {
        // Only log every 5s to reduce noise
        if (timeLeft % 5 === 0) {
          console.log('[ci-queue] Workers are active, waiting for them to complete.');
        }
        timeLeftWithNoWorkers = this.config.inactiveWorkersTimeout;
      } else {
        timeLeftWithNoWorkers -= 1;
      }
    }

    if (timeLeftWithNoWorkers <= 0) {
      console.log('[ci-queue] Aborting, it seems all workers died.');
    }

    const isExhausted = await this.isExhausted();
    if (!isExhausted) {
      console.log('[ci-queue] Not all tests have finished in time');
      return false;
    }

    if (this.config.failureFile) {
      const failedTests = await this.getFailedTests();
      const absolutePath = path.join(process.cwd(), this.config.failureFile);
      const directory = path.dirname(absolutePath);

      mkdirSync(directory, { recursive: true });
      writeFileSync(absolutePath, failedTests);
    }
    return true;
  }

  private async workersAreActive() {
    const zRangeByScoreArr = await this.client.zRangeByScore(
      this.key('running'),
      (Date.now() / 1000) - this.config.timeout,
      '+inf',
      {
        LIMIT: { offset: 0, count: 1 },
      },
    );
    return zRangeByScoreArr.length > 0;
  }
}

File: ruby/lib/ci/queue/redis/supervisor.rb
# frozen_string_literal: true
module CI
  module Queue
    module Redis
      class Supervisor < Base
        def master?
          false
        end

        def total
          wait_for_master(timeout: config.queue_init_timeout)
          redis.get(key('total')).to_i
        end

        def build
          @build ||= CI::Queue::Redis::BuildRecord.new(self, redis, config)
        end

        def wait_for_workers
          wait_for_master(timeout: config.queue_init_timeout)

          yield if block_given?

          time_left = config.report_timeout
          time_left_with_no_workers = config.inactive_workers_timeout
          last_heartbeat_time = Time.now
          until exhausted? || time_left <= 0 || max_test_failed? || time_left_with_no_workers <= 0
            time_left -= 1
            sleep 1

            # Heartbeat log every 5 minutes
            if Time.now - last_heartbeat_time > 300
              puts '[ci-queue] Still working'
              last_heartbeat_time = Time.now
            end

            if active_workers?
              time_left_with_no_workers = config.inactive_workers_timeout
            else
              time_left_with_no_workers -= 1
            end

            yield if block_given?
          end

          puts "Aborting, it seems all workers died." if time_left_with_no_workers <= 0
          exhausted?
        rescue CI::Queue::Redis::LostMaster
          false
        end

        private

        def active_workers?
          # if there are running jobs we assume there are still agents active
          redis.zrangebyscore(key('running'), Time.now.to_f - config.timeout, "+inf", limit: [0,1]).count > 0
        end
      end
    end
  end
end
